### PR TITLE
koordlet: fix RDT cache id issue

### LINKED ISSUE #1777: [BUG] Failed to apply l3 cat and mba policy due to deficiencies in CPU cache info collection.
**What happened**:
```
W1213 10:54:56.286611   27294 resctrl_reconcile.go:253] failed to write l3 cat policy on schemata for group BE, err: failed to read current resctrl schemata, path /host-sys-fs/resctrl/BE/schemata, err: failed to parse l3 schemata, content     L3:0=ffff;1=ffff;2=ffff;3=ffff;4=ffff;5=ffff;6=ffff;7=ffff;16=ffff;17=ffff;18=ffff;19=ffff;20=ffff;21=ffff;22=ffff;23=ffff
    MB:0=2048;1=2048;2=2048;3=2048;4=2048;5=2048;6=2048;7=2048;16=2048;17=2048;18=2048;19=2048;20=2048;21=2048;22=2048;23=2048
, err: read resctrl schemata failed, L3 masks of node 8 is missing
W1213 10:54:56.286629   27294 resctrl_reconcile.go:352] failed to apply l3 cat policy for group BE, err: failed to read current resctrl schemata, path /host-sys-fs/resctrl/BE/schemata, err: failed to parse l3 schemata, content     L3:0=ffff;1=ffff;2=ffff;3=ffff;4=ffff;5=ffff;6=ffff;7=ffff;16=ffff;17=ffff;18=ffff;19=ffff;20=ffff;21=ffff;22=ffff;23=ffff
    MB:0=2048;1=2048;2=2048;3=2048;4=2048;5=2048;6=2048;7=2048;16=2048;17=2048;18=2048;19=2048;20=2048;21=2048;22=2048;23=2048
, err: read resctrl schemata failed, L3 masks of node 8 is missing
W1213 10:54:56.286737   27294 resctrl_reconcile.go:279] failed to write mba policy on schemata for group BE, err: failed to read current resctrl schemata, path /host-sys-fs/resctrl/BE/schemata, err: failed to parse l3 schemata, content     L3:0=ffff;1=ffff;2=ffff;3=ffff;4=ffff;5=ffff;6=ffff;7=ffff;16=ffff;17=ffff;18=ffff;19=ffff;20=ffff;21=ffff;22=ffff;23=ffff
    MB:0=2048;1=2048;2=2048;3=2048;4=2048;5=2048;6=2048;7=2048;16=2048;17=2048;18=2048;19=2048;20=2048;21=2048;22=2048;23=2048
, err: read resctrl schemata failed, L3 masks of node 8 is missing
W1213 10:54:56.286755   27294 resctrl_reconcile.go:356] failed to apply cat MB policy for group BE, err: failed to read current resctrl schemata, path /host-sys-fs/resctrl/BE/schemata, err: failed to parse l3 schemata, content     L3:0=ffff;1=ffff;2=ffff;3=ffff;4=ffff;5=ffff;6=ffff;7=ffff;16=ffff;17=ffff;18=ffff;19=ffff;20=ffff;21=ffff;22=ffff;23=ffff
    MB:0=2048;1=2048;2=2048;3=2048;4=2048;5=2048;6=2048;7=2048;16=2048;17=2048;18=2048;19=2048;20=2048;21=2048;22=2048;23=2048
, err: read resctrl schemata failed, L3 masks of node 8 is missing
```

**What you expected to happen**:
Apply the policy successfully
**How to reproduce it (as minimally and precisely as possible)**:
Enable resctrl in resource-qos-config while the node cache ids are not contiguous.
**Anything else we need to know?**:
I checked the **[Intel and Linux official documentation](https://www.kernel.org/doc/html/v5.4/x86/resctrl_ui.html#cache-ids)** and found that instead of using "socket" or "core" to define the set of logical cpus sharing a resource they use a "Cache ID", and it isn't guaranteed to be a contiguous sequence, there may be gaps. 

But koordinator assumed that the cache IDs are contiguous.[code link](https://github.com/koordinator-sh/koordinator/blob/c927e11a57e105697d43c65e5dfb643adcf096f5/pkg/koordlet/util/system/resctrl.go#L366C6-L366C6)

Koordlet uses `lscpu` command with `-e=CPU,NODE,SOCKET,CORE,CACHE,ONLINE` flag to get the node cpu info, but actually the cache ID info are stored in the `/sys/devices/system/cpu/cpu*/cache/index*/id`
**Environment**:
- App version: v1.3.0
- Kubernetes version (use kubectl version): v1.23.13
- Install details (e.g. helm install args):
- Node environment (for koordlet/runtime-proxy issue):
  - Containerd/Docker version:
  - OS version:
  - Kernal version:
  - Cgroup driver: cgroupfs/systemd
- Others:

### PR DESCRIPTION
**Pull request by bowen-intel**
### Ⅰ. Describe what this PR does
ref: #1777
Temporarily fix the issue of non-contiguous cache IDs by reading the resctrl root schemata. Since it has checked resctrl mount condition, the schemata can be considered reliable. Fix some function name issue.

### Ⅱ. Does this pull request fix one issue?
fixes #1777

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

**Comment by zwzhang0107** [comment]
nice!

File: pkg/koordlet/util/system/resctrl.go [hunk 1]
```diff
@@ -427,6 +441,16 @@ func ReadResctrlSchemataRaw(schemataFile string, l3Num int) (*ResctrlSchemataRaw
 	return schemataRaw, nil
 }
 
+// GetCacheIds get cache ids from schemata file.
+// e.g. schemata=`L3:0=fff;1=fff\nMB:0=100;1=100\n` -> [0, 1]
+func GetCacheIds() ([]int, error) {
```
Comment (zwzhang0107) [inline pkg/koordlet/util/system/resctrl.go:446]: the result can be cached, so that we don't need to read the file every time.

Comment (bowen-intel) [reply pkg/koordlet/util/system/resctrl.go:451]: I am attempting to use `sync.OnceValues` for caching cache IDs. However, this function was introduced in the standard library starting from Go 1.21. Should I temporarily place it in the `util` package and replace the function once Koordinator is upgraded to version 1.21?

**Comment by Rouzip** [comment]
PTAL, thanks 😀 @zwzhang0107

**Review by jasonliu747 — APPROVED** [review]
/lgtm

**Review by hormes — APPROVED** [review]

**Pull request merged**